pkg/middleware/auth.go

```diff
@@ -146,6 +148,49 @@ func Auth(options *AuthOptions) web.Handler {
 	}
 }
 
+func RoleAppPluginAuthAndSignedIn(ps pluginstore.Store) web.Handler {
+	return func(c *contextmodel.ReqContext) {
+		if !c.IsSignedIn {
+			notAuthorized(c)
+			return
+		}
+
+		pluginID := web.Params(c.Req)[":id"]
+		p, exists := ps.Plugin(c.Req.Context(), pluginID)
+		if !exists {
+			c.JsonApiErr(http.StatusNotFound, "Plugin not found", nil)
+			return
+		}
+
+		found := false
+		allowed := false
+		path := filepath.Clean(c.Req.RequestURI)
+		for _, i := range p.Includes {
+			if i.Type != "page" {
+				continue
+			}
+
+			if filepath.Clean(i.Path) == path {
+				found = true
+				if i.Role == "" || c.HasRole(i.Role) {
+					allowed = true
+				}
+				break
+			}
+		}
+
+		if !found {
+			// This isn't an API request, so we probably shouldn't return a JSON error.
+			c.JsonApiErr(http.StatusNotFound, "Plugin page not found", nil)
+			return
+		}
+		if !allowed {
+			accessForbidden(c)
```

Comment:
This will return and skip following RBAC middleware which might be problematic so we probably need to support both RBAC enabled or disabled?

I guess we have this "weird" use case where plugins defines what role a user need to have to access a page where's RBAC allows end users to customize this. What is lacking is plugins being able to define permission/action needed to access the page, but yeah that's some time in the future maybe.

Reply:
Well I guess. I was thinking if someone uses RBAC to define a custom role with allowed access to certain plugin then if it doesn't have the role defined by plugin it will fail :shrug:

Reply:
CC @grafana/identity-access-team for some opinions and 👀

Reply:
Hey, first of thank you for tackling this issue, we had this in the backlog for a while: https://github.com/grafana/grafana-enterprise/issues/4201

Plugin defined access control is still behind the `accessControlOnCall` feature toggle.

We only have OnCall relying on this feature so far. When they define an action or a role to protect a specific frontend route, this is **as a fallback when access control is not enforced on the plugin**
https://github.com/grafana/oncall/blob/53ca5331c669ca98b6eed142b2356f3052f5470b/grafana-plugin/src/plugin.json#L36-L37

So IMO the check should be similar to what we do here:
https://github.com/grafana/grafana/blob/ieva/remove-dash-perm-checks-for-annotations/pkg/services/navtree/navtreeimpl/applinks.go#L238-L252

We first check if we need to enforce RBAC, if that's the case then solely check the user has the action, otherwise, we don't need to enforce RBAC, just check the role.

Reply:
Awesome, thanks for the pointers @gamab. Have shoehorned that in, but now it needs some more test coverage so will give that a stab when I get the chance!